src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,187 @@
+
+            Func<CancellationToken, Task<Document>>? VisitCasePatternSwitchLabel(CasePatternSwitchLabelSyntax node)
+            {
+                var analyzedNode = AnalyzeNode(node, semanticModel, cancellationToken);
+                if (analyzedNode is null)
+                    return null;
+
+                return cancellationToken => document.ReplaceNodeAsync(node, node
+                    .WithPattern(analyzedNode
+                        .AsPatternSyntax(out var expression)
+                        .ConvertToSingleLine()
+                        .WithTriviaFrom(node.Pattern))
+                    .WithWhenClause(expression is null ? null : WhenClause(expression))
+                    .WithAdditionalAnnotations(Formatter.Annotation), cancellationToken);
+            }
+
+            Func<CancellationToken, Task<Document>>? VisitExpression(ExpressionSyntax node)
+            {
+                var expression = GetTopmostExpression(node);
+                if (expression is null)
+                    return null;
+
+                var analyzedNode = AnalyzeNode(expression, semanticModel, cancellationToken);
+                if (analyzedNode is null)
+                    return null;
+
+                return cancellationToken => document.ReplaceNodeAsync(expression, analyzedNode
+                    .AsExpressionSyntax()
+                    .WrapPropertyPatternClauses()
+                    .WithTriviaFrom(expression)
+                    .WithAdditionalAnnotations(Formatter.Annotation), cancellationToken);
+            }
+        }
+
+        private static ExpressionSyntax? GetTopmostExpression(ExpressionSyntax node)
+        {
+            for (SyntaxNode? current = node; current != null; current = current.GetParent(ascendOutOfTrivia: true))
+            {
+                if (!(current is ExpressionSyntax expr))
+                    break;
+
+                expr = expr.WalkUpParentheses();
+                if (expr.IsTopmostExpression())
+                    return expr;
+            }
+
+            return null;
+        }
+
+        private static AnalyzedNode? AnalyzeNode(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            if (node.ContainsDiagnostics)
+                return null;
+
+            var operation = semanticModel.GetOperation(node, cancellationToken);
+            if (operation is null)
+                return null;
+
+            var analyzedNode = AnalyzedNodeFactory.Visit(operation);
+            if (analyzedNode is null)
+                return null;
+
+            var expandedNode = AnalyzedNodeExpander.Expand(analyzedNode);
+            var reducedNode = AnalyzedNodeReducer.Reduce(expandedNode);
+
+            if (reducedNode.Equals(analyzedNode))
+                return null;
+
+            if (reducedNode is OperationEvaluation)
+                return null;
+
+            if (HasIllegalPatternVariables(reducedNode))
+                return null;
```

Comment:
is it possible for expansion/reuction to have multiple results, some which woul dhave illegal variables and some which would not?